packages/mobile/src/identity/commentEncryption.ts

```diff
@@ -0,0 +1,309 @@
+} from 'src/identity/contactMapping'
+import { getUserSelfPhoneHashDetails, PhoneNumberHashDetails } from 'src/identity/privacy'
+import {
+  AddressToE164NumberType,
+  e164NumberToAddressSelector,
+  E164NumberToAddressType,
+  e164NumberToSaltSelector,
+  E164NumberToSaltType,
+} from 'src/identity/reducer'
+import { NewTransactionsInFeedAction } from 'src/transactions/actions'
+import Logger from 'src/utils/Logger'
+import { setPrivateCommentKey } from 'src/web3/actions'
+import { getContractKit } from 'src/web3/contracts'
+import { privateCommentKeySelector } from 'src/web3/selectors'
+
+const TAG = 'identity/commentKey'
+// A separator to split the comment content from the metadata
+const METADATA_CONTENT_SEPARATOR = '~'
+// Format should be separator + e164Number + salt
+const PHONE_METADATA_REGEX = new RegExp(
+  `(.*)${METADATA_CONTENT_SEPARATOR}([+][1-9][0-9]{1,14})([a-zA-Z0-9+/]{13})$`
+)
+
+// Derive a new comment key from the provided private key
+export function* createCommentKey(seedPrivateKey: string) {
+  Logger.debug(TAG, 'Creating a new comment key')
+  const privateCEK = bufferToHex(deriveCEK(seedPrivateKey))
+  yield put(setPrivateCommentKey(privateCEK))
+}
+
+export function* getCommentKey(address: string) {
+  const contractKit = yield call(getContractKit)
+  const accountsWrapper: AccountsWrapper = yield call([
+    contractKit.contracts,
+    contractKit.contracts.getAccounts,
+  ])
+  const hexString: string = yield call(accountsWrapper.getDataEncryptionKey, address)
+  return !hexString ? null : hexToBuffer(hexString)
+}
+
+export function* encryptComment(
+  comment: string | null,
+  toAddress: string | null,
+  fromAddress: string | null,
+  includePhoneNumMetadata: boolean = false
+) {
+  Logger.debug(TAG + 'encryptComment', 'Encrypting comment')
+  if (!features.USE_COMMENT_ENCRYPTION || !comment || !toAddress || !fromAddress) {
+    Logger.debug(TAG, 'Invalid params, skipping encryption')
+    return comment
+  }
+
+  // TODO currently users register this key when the get verified
+  // We should nudge unverified users to register a key as well otherwise
+  // they don't benefit from comment encryption
+  const fromKey: Buffer | null = yield call(getCommentKey, fromAddress)
+  if (!fromKey) {
+    Logger.debug(TAG + 'encryptComment', 'No sender key found, skipping encryption')
+    return comment
+  }
+
+  const toKey: Buffer | null = yield call(getCommentKey, toAddress)
+  if (!toKey) {
+    Logger.debug(TAG + 'encryptComment', 'No recipient key found, skipping encryption')
+    return comment
+  }
+
+  let commentToEncrypt = comment
+  if (features.PHONE_NUM_METADATA_IN_TRANSFERS && includePhoneNumMetadata) {
+    Logger.debug(TAG + 'encryptComment', 'Including phone number metadata in comment')
+    const selfPhoneDetails: PhoneNumberHashDetails | undefined = yield call(
+      getUserSelfPhoneHashDetails
+    )
+    commentToEncrypt = embedPhoneNumberMetadata(comment, selfPhoneDetails)
```

Comment:
Out of curiosity, why is embedding phone number metadata necessary?

Reply:
We used to lookup a user's whole address book on-chain to see who's verified. It was slow but it worked. Now that's not possible because of PGPNP. So we include the metadata in comments, enabling the recipient to connect the dots